src/cosl/coordinated_workers/interface.py

```diff
@@ -403,6 +406,29 @@ def get_address_from_role(self, role: str) -> Optional[str]:
             return address_set.pop()
         return None
 
+    def _has_worker_published(self, relation: ops.Relation) -> bool:
+        """Verify that the worker has published its data.
+
+        - unit address is published
+        - roles are published
+        """
+        if not relation.app:
+            return False
+        if len(relation.units) == 0:
+            return False
+        for worker_unit in relation.units:
+            try:
+                ClusterRequirerUnitData.load(relation.data[worker_unit])
+                if self._charm.unit.is_leader():
+                    ClusterRequirerAppData.load(relation.data[relation.app])
+            except DataValidationError as e:
+                log.info(
```

Comment:
I'm pretty sure we're already logging this somewhere else at debug level, can you double check that this won't happen twice per hook?

Reply:
besides, the point of this method is to check whether the remote end has published valid data. a true/false outcome is enough, I don't think we should be doing any more than that.